File: scripts/gerar_aulas_dadas.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Script para gerar CSV de aulas dadas com dados fictícios
para todas as escolas do arquivo escolas.csv
"""

import csv
import random
from datetime import datetime, timedelta
from pathlib import Path

def get_dias_uteis(ano, mes):
    """Retorna lista de dias úteis (segunda a sexta) de um mês"""
    dias = []
    # Último dia do mês
    if mes == 12:
        ultimo_dia = 31
    else:
        ultimo_dia = (datetime(ano, mes + 1, 1) - timedelta(days=1)).day
    
    for dia in range(1, ultimo_dia + 1):
        data = datetime(ano, mes, dia)
        # 0 = segunda, 4 = sexta
        if data.weekday() < 5:  # Segunda a sexta
            dias.append(data)
    
    return dias

def formatar_data(data):
    """Formata data como YYYY-MM-DD"""
    return data.strftime('%Y-%m-%d')

def carregar_escolas(caminho_csv):
    """Carrega escolas do CSV"""
    escolas = []
    
    with open(caminho_csv, 'r', encoding='utf-8') as f:
        reader = csv.DictReader(f)
        for row in reader:
            # Determinar turmas baseado no tipo de ensino
            turmas = []
            
            # Ensino Fundamental: 1º ao 5º ano
            if row.get('ensino_fundamental', '').lower() == 'true':
                # Adicionar séries do 1º ao 5º ano
                turmas.extend(['1º Ano', '2º Ano', '3º Ano', '4º Ano', '5º Ano'])
            
            # Se tiver turmas, adicionar escola
            if turmas:
                escolas.append({
                    'id': int(row['id']),
                    'nome': row['nome'],
                    'municipio': row['municipio'],
                    'regional': row['regional'],
                    'turmas': turmas
                })
    
    return escolas

def gerar_csv_aulas_dadas():
    """Gera CSV de aulas dadas"""
    # Caminhos dos arquivos
    script_dir = Path(__file__).resolve().parent
    base_dir = script_dir.parent
    escolas_csv = base_dir / 'public' / 'escolas.csv'
    output_public = base_dir / 'public' / 'aulas-dadas.csv'
    output_data = base_dir / 'data' / 'aulas-dadas.csv'
    
    print(f"Script dir: {script_dir}")
    print(f"Base dir: {base_dir}")
    print(f"Escolas CSV: {escolas_csv}")
    print(f"Existe escolas.csv? {escolas_csv.exists()}")
    
    # Carregar escolas
    print(f"Carregando escolas de {escolas_csv}...")
    escolas = carregar_escolas(escolas_csv)
    print(f"Total de escolas encontradas: {len(escolas)}")
    
    # Obter todos os dias úteis de fev, mar e abr 2025
    dias_fev = get_dias_uteis(2025, 2)
    dias_mar = get_dias_uteis(2025, 3)
    dias_abr = get_dias_uteis(2025, 4)
    todos_dias = dias_fev + dias_mar + dias_abr
    
    print(f"Total de dias úteis: {len(todos_dias)}")
    print(f"  - Fevereiro: {len(dias_fev)} dias")
    print(f"  - Março: {len(dias_mar)} dias")
    print(f"  - Abril: {len(dias_abr)} dias")
    
    # Gerar linhas do CSV
    linhas = [['escola_id', 'escola_nome', 'regional', 'municipio', 'turma', 'data', 'dia_letivo', 'aulas_dadas']]
    
    total_linhas = 0
    
    for escola in escolas:
        for turma in escola['turmas']:
            dia_letivo = 1
            for data in todos_dias:
                # Gerar número aleatório de aulas dadas (entre 4 e 6 aulas por dia)
                aulas_dadas = random.randint(4, 6)
                
                linha = [
                    escola['id'],
                    escola['nome'],
                    escola['regional'],
                    escola['municipio'],
                    turma,
                    formatar_data(data),
                    dia_letivo,
                    aulas_dadas
                ]
                
                linhas.append(linha)
                dia_letivo += 1
                total_linhas += 1
    
    # Escrever arquivos
    print(f"\nGerando CSV...")
    print(f"Total de linhas (sem header): {total_linhas}")
    
    # Escrever em public/aulas-dadas.csv
    with open(output_public, 'w', encoding='utf-8', newline='') as f:
        writer = csv.writer(f)
        writer.writerows(linhas)
    print(f"✓ Arquivo criado: {output_public}")
    
    # Escrever em data/aulas-dadas.csv
    with open(output_data, 'w', encoding='utf-8', newline='') as f:
        writer = csv.writer(f)
        writer.writerows(linhas)
    print(f"✓ Arquivo criado: {output_data}")
    
    print(f"\n✓ CSV gerado com sucesso!")
    print(f"  - Total de escolas: {len(escolas)}")
    print(f"  - Total de turmas: {sum(len(e['turmas']) for e in escolas)}")
    print(f"  - Total de registros: {total_linhas}")

if __name__ == '__main__':
    gerar_csv_aulas_dadas()

File: scripts/gerar-aulas-dadas.js
// Script para gerar CSV de aulas dadas de forma dinâmica
import fs from 'fs';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

// Função para obter dias úteis de um mês
function getDiasUteis(ano, mes) {
  const dias = [];
  const ultimoDia = new Date(ano, mes, 0).getDate();
  
  for (let dia = 1; dia <= ultimoDia; dia++) {
    const data = new Date(ano, mes - 1, dia);
    const diaSemana = data.getDay();
    // 0 = Domingo, 6 = Sábado
    if (diaSemana > 0 && diaSemana < 6) {
      dias.push(data);
    }
  }
  
  return dias;
}

// Função para formatar data como YYYY-MM-DD
function formatarData(data) {
  const ano = data.getFullYear();
  const mes = String(data.getMonth() + 1).padStart(2, '0');
  const dia = String(data.getDate()).padStart(2, '0');
  return `${ano}-${mes}-${dia}`;
}

// Carregar escolas do CSV
async function carregarEscolas() {
  try {
    const csvPath = join(__dirname, '..', 'public', 'escolas.csv');
    const csvContent = fs.readFileSync(csvPath, 'utf8');
    const linhas = csvContent.split('\n');
    const headers = linhas[0].split(',');
    
    const escolas = [];
    for (let i = 1; i < linhas.length; i++) {
      if (!linhas[i].trim()) continue;
      
      const valores = linhas[i].split(',');
      const escola = {};
      headers.forEach((header, index) => {
        escola[header.trim()] = valores[index]?.trim() || '';
      });
      
      // Determinar turmas baseado no tipo de ensino
      const turmas = [];
      if (escola.ensino_fundamental === 'true') {
        turmas.push('6º Ano', '7º Ano', '8º Ano', '9º Ano');
      }
      if (escola.ensino_medio === 'true') {
        turmas.push('1ª Série', '2ª Série', '3ª Série');
      }
      
      escolas.push({
        id: parseInt(escola.id),
        nome: escola.nome,
        municipio: escola.municipio,
        regional: escola.regional,
        turmas: turmas
      });
    }
    
    return escolas;
  } catch (error) {
    console.error('Erro ao carregar escolas:', error);
    return [];
  }
}

// Gerar CSV
async function gerarCSV() {
  const escolas = await carregarEscolas();
  
  // Obter todos os dias úteis de fev, mar e abr 2025
  const diasFev = getDiasUteis(2025, 2);
  const diasMar = getDiasUteis(2025, 3);
  const diasAbr = getDiasUteis(2025, 4);
  const todosDias = [...diasFev, ...diasMar, ...diasAbr];
  
  const linhas = ['escola_id,escola_nome,regional,municipio,turma,data,dia_letivo,aulas_dadas'];
  
  escolas.forEach(escola => {
    escola.turmas.forEach(turma => {
      let diaLetivo = 1;
      todosDias.forEach((data) => {
        // Gerar número aleatório de aulas dadas (entre 4 e 6 aulas por dia)
        const aulasDadas = Math.floor(Math.random() * 3) + 4; // 4, 5 ou 6
        
        const linha = [
          escola.id,
          `"${escola.nome}"`,
          `"${escola.regional}"`,
          `"${escola.municipio}"`,
          `"${turma}"`,
          formatarData(data),
          diaLetivo++,
          aulasDadas
        ].join(',');
        
        linhas.push(linha);
      });
    });
  });
  
  const csvContent = linhas.join('\n');
  const publicPath = join(__dirname, '..', 'public', 'aulas-dadas.csv');
  const dataPath = join(__dirname, '..', 'data', 'aulas-dadas.csv');
  
  fs.writeFileSync(publicPath, csvContent, 'utf8');
  fs.writeFileSync(dataPath, csvContent, 'utf8');
  
  console.log(`CSV gerado com sucesso!`);
  console.log(`Total de linhas: ${linhas.length - 1}`);
  console.log(`Arquivos criados:`);
  console.log(`  - ${publicPath}`);
  console.log(`  - ${dataPath}`);
}

gerarCSV().catch(console.error);

